### PR TITLE
[rewrite] The overview/state-scripts section

### PR DESCRIPTION
**Pull request by oleorhagen**
The state script section was mostly pretty good. Only minor cosmetics changed
and added.

Changelog: None

Signed-off-by: Ole Petter <ole.orhagen@northern.tech>

### CODE REVIEW

File: 02.Overview/08.State-scripts/docs.md [hunk 1]
```diff
@@ -41,101 +41,149 @@ If Mender is used in standalone mode (installing via command line), some states
...
+* **Artifact scripts**: These are scripts which are added as a part of an update
+  to the Artifact. Scripts can thereby customize actions of the Mender state
+  machine, as described [above](The Mender client states supporting state-script
+  customization) during an update installation. Artifact scripts are easy to
+  distinguish from root file system scripts, as they are prefixed with
+  `Artifact`.
+  
+The reason for having both `root file system` and `Artifact` scripts is due to
+the need that some scripts must run prior to the download of the Artifact and as
+such are notdelivered with the Artifact. These scripts are `Idle`, `Sync` and
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:66]: `notdelivered` -> `not delivered`

File: 02.Overview/08.State-scripts/docs.md [hunk 2]
```diff
@@ -41,101 +41,149 @@ If Mender is used in standalone mode (installing via command line), some states
...
+```
+
+1) The state in which the script executes. 
+2) Execute it before or after entering the state. 
+3) If multiple scripts exist for this state-transition, this decides the script ordering.
+4) Optional description of script functionality. This has no effect on the execution of the scripts.
+
+!!! Note: `Idle`, `Sync`, `ArtifactRollback`, `ArtifactRollbackReboot` ignore
+errors. This means that no error scripts are executed for errors in these
+states.
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:113]: > This means that no error scripts are executed for errors in these states.

What are error scripts?

File: 02.Overview/08.State-scripts/docs.md [hunk 3]
```diff
@@ -41,101 +41,149 @@ If Mender is used in standalone mode (installing via command line), some states
...
 [StateScriptRetryIntervalSeconds](../../client-configuration/configuration-file/configuration-options/#statescriptretryintervalseconds)
-before the state script is called again. Note that scripts are not allowed to
-retry for infinitely long. Please see description of
+before retrying the state script. Scripts are not allowed to retry for
+infinitely long, a total retry window is given to each script. A script retrying
+for longer than
 [StateScriptRetryTimeoutSeconds](../../client-configuration/configuration-file/configuration-options/#statescriptretrytimeoutseconds)
-for more information.
+counts as an error. Only the **Reboot_Enter** state transition, can be
+prolonged indefinitely.
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:137]: > Only the **Reboot_Enter** state transition, can be prolonged indefinitely.

Can it? It does not care about `StateScriptRetryTimeoutSeconds`?

File: 02.Overview/08.State-scripts/docs.md [hunk 4]
```diff
@@ -41,101 +41,149 @@ If Mender is used in standalone mode (installing via command line), some states
...
-aborted and rolled back. In addition, return code `21` is used for the
-[retry-later](#retry-later) feature. All other return codes are reserved for
-future use by Mender and should not be used.
+* Success - 0
+
+Successful script execution - Continue the update.
+
+* Error - 1
+
+Abort update and roll back
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:126]: `Mender client marks the update as failed and attempts to perform a rollback`

File: 02.Overview/08.State-scripts/docs.md [hunk 5]
```diff
@@ -41,101 +41,149 @@ If Mender is used in standalone mode (installing via command line), some states
...
+dealt with properly. The Mender client has built in protections for handling
+such an event. If you are using state scripts this adds some complexity to the
+logic of your scripts. Take this into account when writing a custom state
+script. Follow the below rules when writing your state-scripts, to make sure no
+unintended side-effects are encountered:
+
+### Full rootfs updates
+
+* A power loss will return the Mender state-machine to the `ArtifactFailure** state upon reboot.
+* A power loss happens inside one of the error states will cause the state to repeat until it succeeds without a power-interruption.
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:165]: > Follow the below rules when writing your state-scripts, to make sure no unintended side-effects are encountered:

This entry to the bullet list reads strange to me.

> A power loss happens inside one of the error states will

That?

> the error states will cause the state to repeat until it succeeds without a power-interruption.

Maybe I missed this or needs some additional clarification, which are the error states?

File: 02.Overview/08.State-scripts/docs.md [hunk 6]
```diff
@@ -41,101 +41,149 @@ If Mender is used in standalone mode (installing via command line), some states
...
+** `ArtifactReboot_Enter** state: A power-loss is not an error.
+** `ArtifactCommit_Leave` state: A power-loss will cause this state to rerun, since after a commit, it is too late to do a rollback.
+
+### Partial updates
+
+* A power loss will return the Mender state-machine to the `ArtifactRollback** state upon reboot.
+* A power loss happens inside one of the error states will cause the state to repeat until it succeeds without a power-interruption.
+* Exceptions:
+** `ArtifactReboot_Enter** state: A power-loss is not an error.
+** `ArtifactCommit_Leave` state: A power-loss will cause this state to rerun, since after a commit, it is too late to do a rollback.
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:176]: What is the difference between Full rootfs updates and this?

File: 02.Overview/08.State-scripts/docs.md [hunk 7]
```diff
@@ -41,101 +41,149 @@ If Mender is used in standalone mode (installing via command line), some states
...
+!!! state forever, and thus an install is only installed on a power-cycle of
 !!! the device.
 
 Because of the possible re-execution described above, state scripts should be
-written to be idempotent. This means that re-running the script several times,
-even partially, should have the same effect as running it once, as long as the
-last execution is a complete one.
+idempotent. This means that re-running the script several times, even partially,
+should have the same effect as running it once, as long as the last execution is
+a complete.
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:186]: `a complete` -> `complete`

File: 02.Overview/08.State-scripts/docs.md [hunk 8]
```diff
@@ -145,36 +193,36 @@ Mender deployment log, so it becomes available [locally on the
 client](../../troubleshooting/mender-client#deployment-log-files) as well as
 reported to the server (if the deployment fails) to ease diagnostics.
 
-Thus the state scripts should be written so they output diagnostics
-information to standard error, especially in case of failure
-(returning 1). The maximum size of the log is 10KiB per state script,
-anything above this volume will be truncated.
+Thus, write state scripts to output diagnostics information to standard error,
+especially in case of a failure (exit code 1). Note that the maximum size of the
+log is 10KiB per state script, anything above this size is truncated.
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:198]: I really though that we captured `stdout` as well. At least based on my experience with modules so far.

File: 02.Overview/08.State-scripts/docs.md [hunk 9]
```diff
@@ -200,7 +248,7 @@ to enable this use case.
 !! Maximum wait time between `Sync` and `Download` state is 24 hours, after this
 !! period the update will be marked as failed by the Mender client. This happens
 !! because the Mender Artifact download link is generated in `Sync` state and it
-!! has a expiration time (24 hours).
+!! has an expiration time.
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:251]: Why remove the expiration time? This immediately triggers a question of "How long is the expiration time?" :)

File: 02.Overview/08.State-scripts/docs.md [hunk 10]
```diff
@@ -224,25 +272,24 @@ default on the device, so you want to selectively enable it when needed.
...
 
 
 ## Including state scripts in Artifacts and disk images
 
-The easiest way to have Mender run the state scripts is to create a new
-OpenEmbedded recipe that inherits `mender-state-scripts` and copies them into
-place in `do_compile`, using the
+The easiest way to have Mender run state scripts is to create a new OpenEmbedded
+recipe that inherits `mender-state-scripts` and copies them into place in
+`do_compile`, using the
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:292]: I would remove this section unless we have examples of the following:

1. Create Artifact with state-script using `mender-artifact`
2. Add state-scripts using mender-convert
3. Add state-scripts using Yocto

Though I think above belong to their respective section in the docs. So remove or link :)

Comment (oleorhagen) [reply 02.Overview/08.State-scripts/docs.md:292]: Removing sections is always welcome :)

File: 02.Overview/08.State-scripts/docs.md (hunk 8 above)
Comment (oleorhagen) [reply 02.Overview/08.State-scripts/docs.md:198]: Curious. It only reads the std-err pipe in the code at least

Comment (mirzak) [reply 02.Overview/08.State-scripts/docs.md:198]: Quick test shows that you are correct :)

Comment (oleorhagen) [reply 02.Overview/08.State-scripts/docs.md:198]: :)

File: 02.Overview/08.State-scripts/docs.md [hunk 11]
```diff
@@ -5,34 +5,34 @@ taxonomy:
...
-* **ArtifactRollback**: if the new update is broken and we need to go back to the previous one
-* **ArtifactRollbackReboot**: if we need to reboot the device after performing rollback
-* **ArtifactFailure**: if any of the "Artifact" states are failing, the device enters and executes this state. This state always runs after the ArtifactRollback and ArtifactRollbackReboot states.
+### Managed mode
 
-State scripts can either be run as we transition into a state; "Enter", or out from a state, "Leave". Most of the states also have an "Error" transition which is run when some error occurs while executing any action inside the given state (including execution of Enter and Leave scripts).
+* **Idle**: The Mender client is idling, waiting for the next action to handle. At this stage, no communication with the server, or downloads are in progress.
+* **Sync**: At this stage the Mender client will either send or update its inventory to the server, or check if an update is available. This requires communication with the server.
+* **Download**: When an update is ready at the server side Mender downloads it (i.e. streams) to the inactive rootfs partition.
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:22]: (i.e. streams it) ? I am not sure but it sounds strange without the "it" there.

File: 02.Overview/08.State-scripts/docs.md [hunk 12]
```diff
@@ -5,34 +5,34 @@ taxonomy:
...
+* **ArtifactInstall**: Swap the active and inactive partitions through interaction with the bootloader.
+* **ArtifactReboot**: Reboots the device. Note: the Enter scripts are run prior to reboot; the Leave actions are run after.
+* **ArtifactCommit**: If the new image in the passive partition passes all the integrity checks, the Mender client will mark the update as successful and continue running from this partition. The commit makes the update persistent. A pre-script here can add custom integrity checks beyond what the Mender client provides out of the box.
+* **ArtifactRollback**: if the new update fails any of the above mentioned integrity checks, the Mender client will revert the update payload. At this point, scripts in this state can help revert migrations done as a part of the update..
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:26]: If (capital i)

Comment (lluiscampos) [reply 02.Overview/08.State-scripts/docs.md:26]: Extra dot at "update.."

File: 02.Overview/08.State-scripts/docs.md [hunk 13]
```diff
@@ -5,34 +5,34 @@ taxonomy:
...
+* **ArtifactRollbackReboot**: Some update types - most notably rootfs images - if any of the integrity checks fail, require a reboot. This state runs right before the client reboots back into the old image partition.
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:27]: Suggestion: "the client reboots back (i.e. rollback) into the..."

File: 02.Overview/08.State-scripts/docs.md [hunk 14]
```diff
@@ -5,34 +5,34 @@ taxonomy:
...
+* **ArtifactFailure**: if any of the States with an "Artifact" prefix fail, execute this state. Note: this state always runs after the ArtifactRollback and ArtifactRollbackReboot states.
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:28]: If

Comment (lluiscampos) [reply 02.Overview/08.State-scripts/docs.md:28]: "after the ArtifactRollback and the ArtifactRollbackReboot states" or "after ArtifactRollback and ArtifactRollbackReboot states".

File: 02.Overview/08.State-scripts/docs.md [hunk 15]
```diff
@@ -41,101 +41,141 @@ If Mender is used in standalone mode (installing via command line), some states
...
-will take over.
+There are two types of the state scripts:
+
+* **Root file system scripts**: These are not a part of an update, but scripts
+which are already present on a device, stored as a part of the current root file
+system. The default location is `/etc/mender/scripts`.
+
+* **Artifact scripts**: These are scripts which are added as a part of an update
+  to the Artifact. Scripts can thereby customize actions of the Mender state
+  machine, as described [above](The Mender client states supporting state-script
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:59]: Missing `#` in the link

File: 02.Overview/08.State-scripts/docs.md [hunk 16]
```diff
@@ -41,101 +41,141 @@ If Mender is used in standalone mode (installing via command line), some states
...
+An example of which is:
+
+```
+Idle_Enter_01_disable-network
+```
+
+and
+
+```
+Idle_Enter_01_enable-network
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:93]: Leave?

File: 02.Overview/08.State-scripts/docs.md [hunk 17]
```diff
@@ -41,101 +41,141 @@ If Mender is used in standalone mode (installing via command line), some states
...
+  `Artifact`.
+  
+The reason for having both `root file system` and `Artifact` scripts is due to
+the need that some scripts must run prior to the download of the Artifact and as
+such are not delivered with the Artifact. These scripts are `Idle`, `Sync` and
+`Download`. An example of such a use-case is disabling network connectivity upon
+entering `Idle`, and re-enabling it upon leaving `Idle`. This distinction is
+important to remember, as when deploying a new update all scripts that run up
+until - and not including - the `ArtifactInstall` state are from the old
+Artifact installation - not the new one.
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:71]: This sentence made me confused. Does it mean than after installing an artifact with "Artifact scripts" on it, these are installed as "Root file system scripts" for consecutive installations? I know that it does not as that they are one off, but we should try to rephrase this.

File: 02.Overview/08.State-scripts/docs.md [hunk 18]
```diff
@@ -41,101 +41,141 @@ If Mender is used in standalone mode (installing via command line), some states
...
-[retry-later](#retry-later) feature. All other return codes are reserved for
-future use by Mender and should not be used.
+* Success - 0
 
-### Retry-later
+Successful script execution - Continue the update.
 
-State scripts are allowed to return a specific error code (`21`), in which case
-the client will sleep for a time configured by
+* Error - 1
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:127]: The first read  looked like the script should return `-1`. Suggestions: `Error (1)` or `Error: 1`

File: 02.Overview/08.State-scripts/docs.md [hunk 19]
```diff
@@ -41,101 +41,141 @@ If Mender is used in standalone mode (installing via command line), some states
...
+
+Mender client marks the update as failed and attempts to perform a rollback.
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:129]: It depends on the state, right? I would say "abort the update or perform a rollback"

File: 02.Overview/08.State-scripts/docs.md [hunk 20]
```diff
@@ -41,101 +41,141 @@ If Mender is used in standalone mode (installing via command line), some states
...
+!!! All other return codes are reserved for future use by Mender and should not be used.
+
 
 ## Script timeout
 
 Each script has a maximum execution time defined by
-[StateScriptTimeoutSeconds](../../client-configuration/configuration-file/configuration-options/#statescripttimeoutseconds).
-If a script exceeds this running time, its process group will be killed and the
-Mender client will treat the script and the update as failed.
+[StateScriptTimeoutSeconds](../../client-configuration/configuration-file/configuration-options/#statescripttimeoutseconds**.
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:151]: Remove `**`

File: 02.Overview/08.State-scripts/docs.md [hunk 21]
```diff
@@ -41,101 +41,141 @@ If Mender is used in standalone mode (installing via command line), some states
...
 
-In general power loss means that Mender will transition into an error state,
-either ArtifactRollback or ArtifactFailure, depending on the level of rollback
-support of the Payload types in the Artifact. If a power loss happens inside one
-of the error states, that state will be repeated until it succeeds without a
-power interruption. However, there are some exceptions: Power loss is allowed to
-happen within any Reboot state, since it may be indistinguishable from a normal
-reboot. Also `ArtifactCommit_Leave` will be repeated just like the error states,
-since after committing it is too late to do a rollback.
+A power loss during an update is a catastrophic event for an IoT device if not
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:157]: Suggestion: "can be a catastrophic..."

File: 02.Overview/08.State-scripts/docs.md [hunk 22]
```diff
@@ -41,101 +41,141 @@ If Mender is used in standalone mode (installing via command line), some states
...
+dealt with properly. The Mender client has built in protections for handling
+such an event. If you are using state scripts this adds some complexity to the
+logic of your scripts. Take this into account when writing a custom state
+script. Follow the below rules when writing your state-scripts, to make sure no
+unintended side-effects are encountered:
+
+* A power loss will return the Mender state-machine to the `ArtifactFailure** state upon reboot.
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:164]: Remove `**`, missing `

File: 02.Overview/08.State-scripts/docs.md [hunk 23]
```diff
@@ -41,101 +41,141 @@ If Mender is used in standalone mode (installing via command line), some states
...
+* A power loss when executing an error state (i.e., ArtifactFailure, ArtifactRollback, ArtifactRollbackReboot) will cause it to be repeated.
+* Exceptions:
+** `ArtifactReboot_Enter** state: A power-loss is not an error.
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:167]: Missing `

File: 02.Overview/08.State-scripts/docs.md [hunk 24]
```diff
@@ -145,36 +185,36 @@ Mender deployment log, so it becomes available [locally on the
...
-anything above this volume will be truncated.
+Thus, write state scripts to output diagnostics information to standard error,
+especially in case of a failure (exit code 1). Note that the maximum size of the
+log is 10KiB per state script, anything above this size is truncated.
 
 ## Example use cases
 
-Mender users will probably come up with a lot of interesting use cases for state
-scripts, and we will cover some well-known ones below for inspiration.
+It is up to you to yield the state-scripts to suit your needs, and your fantasy
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:194]: creativity? I think fantasy sounds non technical.

File: 02.Overview/08.State-scripts/docs.md [hunk 25]
```diff
@@ -145,36 +185,36 @@ Mender deployment log, so it becomes available [locally on the
...
+is the limitation. Mender state-scripts are written to put as few limitations on
```
Comment (lluiscampos) [inline 02.Overview/08.State-scripts/docs.md:195]: Suggestion: "Mender state-scripts feature is design as per to put as few...", as the state scripts are written by the user, if any.

File: 02.Overview/08.State-scripts/docs.md [hunk 26]
```diff
@@ -5,34 +5,34 @@ taxonomy:
...
 ---
 
-The Mender Client has the ability to run pre- and postinstall scripts, before
-and after it writes the root file system. However, Mender state scripts are more
-general and useful than pre/postinstall scripts because they can be run between
-*any* state transition, not just (before/after) the install state. For some
-examples of usage, see [example use cases](#example-use-cases).
+State scripts offers the opportunity to customize device operability during an
+update installation. The Mender Client can run pre- and post-install scripts,
+before and after every state in the Mender state-machine. For some examples of
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply 02.Overview/08.State-scripts/docs.md:10]: Lot's easier. thx

**Pull request merged**